[Additional Fields] Fix persistence and partial updates of additional fields

alexflorisca opened the pull request:
<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

This PR fixes 4 bugs with the additional fields on checkout and some flakey end to end tests

1. The `woocommerce_store_api_checkout_update_order_from_request` action is not being called on the \PUT endpoint of the checkout.

There was some duplication of the code in `update_order_from_request` function used by the checkout POST route. I've refactored the checkout /PUT route to use the same function and therefore call the `woocommerce_store_api_checkout_update_order_from_request` action.

Bug introduced in PR #53553.

2. Additional fields registered in the contact location were not being persisted.

Needed to get fields for the `contact` location as well as `order` when persisting.

Bug introduced in PR #55570.

3. Required checkboxes persist invalid values

With the work in #53553, when a field has an invalid value, it still gets persisted. With text fields this isn't an issue because the validation is done server side and if they are empty they do not get persisted. But with checkboxes, this isn't the case, and the invalid value was being persisted. This PR addresses that.

4. [Partial address overwritten when other fields are persisted](https://github.com/woocommerce/woocommerce/issues/56155)

Fixes #56155.

When a shopper visits the checkout block, enters a partial address, then changes another additional field, order notes or payment method, the address on the client is overwritten with a blank address from the server. This happens because the address (both shipping and billing) aren't persisted until a full address is entered.

The fix for this was to use `receiveCartContents` rather than `receiveCart` which updates the cart, but not the addresses. We also added calls to `__internalIncrementCalculating()` and `__internalDecrementCalculating` in order for `waitForCheckoutToFinishUpdating()` to return true if the PUT request is still happening. This is used in the checkout e2e tests.

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Add the code from [this gist](https://gist.github.com/alexflorisca/051e0fe9b0bae22f5c4efdae9c505ba9) to your functions.php, woocommerce.php or add it using a code snippet plugin.
2. Install the [Additional Field Tester plugin](https://github.com/woocommerce/additional-checkout-fields-tester). 
3. Insert an additional required checkbox either with a code snippet plugin or by adding this code to `woocommerce.php` or `functions.php` files:

```
woocommerce_register_additional_checkout_field(
		array(
			'id'       => 'first-plugin-namespace/truck-size-ok',
			'label'    => 'Can a truck fit down your road?',
			'location' => 'address',
			'type'     => 'checkbox',
                         'required' => true,
		)
	);
```

6. Enable the Cash on delivery, cheque and bank transfer payment methods.
7. Add some products to your Cart and visit the Checkout. 
8. Change between the above 3 payment methods, and check that the totals update to add 2, 3 and 5 units of currency to the total.
9. Edit each of the additional fields and wait for a /PUT request to checkout endpoint
10. Refresh the page
11. Check that the value is persisted.
12. Click the required checkbox. There should be a PUT request to the checkout endpoint.
13. Untick the checkbox. There should be no request and you should see an inline error.

**Test that the address doesn't get overridden by the PUT request**
1. Open a new incognito window, add some items to cart and visit the checkout page
2. Fill in the address partially, miss out the postcode.
3. Fill in another additional field and check that a PUT request is successful.
4. The address you entered should still be there.
5. Enter the postcode and any other missing required fields.
6. Check for a request to the /batch endpoint
7. Modify one of the additional fields again.
8. Check that the address is not overwritten.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [x] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [x] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [x] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

Fix contact additional fields not persisting, required checkboxes persisting when invalid and `woocommerce_store_api_checkout_update_order_from_request` not triggered on a PUT request to the `/wc/store/v1/checkout` endpoint.

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

Commit:
add contact fields to the persisting fields function

Commit:
Remove duplicate code and use update_order_from_request for PUT requests also

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

Commit:
Validate additional fields before pushing

Commit:
Add changefile(s) from automation for the following project(s): woocommerce

mikejolley reviewed (commented):
I like this approach @alexflorisca I followed your testing steps and tested with 2 required fields, one being a checkbox.

- When the checkbox is invalid, no push ✅ 
- When the checkbox was valid, push ✅ 
- When the checkbox is in an invalid state and I edit the other field, pushes the valid field ✅ 

So I think this looks good.

I do think we need to insert a unit/e2e (?) test to check this functionality—hopefully GPT can write something to test the above case? That way we can avoid another regression.

mikejolley commented on plugins/woocommerce-blocks/assets/js/data/checkout/utils.ts:
```diff
@@ -230,3 +233,36 @@ export const getPaymentResultFromCheckoutResponse = (
+	// The validation store prefixes fields with 'additional-field-'
+	for ( const fieldKey of Object.keys( additionalFields ) ) {
+		// We don't know which location the field is in, so we need to check both
+		let error = validationStore.getValidationError(
```
@senadir is there a better way to detect or know where a field is on client side? I feel like there will be.

mikejolley replied:
@alexflorisca Nadir shared this which can be reused instead of checking for both https://github.com/woocommerce/woocommerce/blob/74c589d6efea4003652e6b1d3eb40291549c9001/plugins/woocommerce-blocks/assets/js/settings/blocks/constants.ts#L124-L155

Commit:
re-use constants to check for field prefix

Commit:
e2e test to for persisting fields

alexflorisca commented:
Thanks Mike, just pushed some changes - a test and using the constants file you pointed out. Have another look when you can 👍

mikejolley commented on plugins/woocommerce-blocks/tests/e2e/tests/checkout/additional-fields.shopper.block_theme.spec.ts:
```diff
@@ -908,5 +908,101 @@ test.describe( 'Shopper → Additional Checkout Fields', () => {
 			);
 		} );
+
+		test( 'Shopper can refresh the page, and not loose any entered data', async ( {
```
```suggestion
		test( 'Shopper can refresh the page, and not lose any entered data', async ( {
```

mikejolley commented on plugins/woocommerce-blocks/tests/e2e/tests/checkout/additional-fields.shopper.block_theme.spec.ts:
```diff
@@ -908,5 +908,101 @@ test.describe( 'Shopper → Additional Checkout Fields', () => {
+				)
+				.check();
+
+			// Check that contact and order fields are persisted
+			await checkoutPageObject.page.waitForRequest(
+				'**/wc/store/v1/checkout**'
+			);
+
+			// Check that address fields are persisted
+			await checkoutPageObject.page.waitForRequest(
+				'**/wc/store/v1/batch**'
+			);
```
Is it normal for both to be triggered?

alexflorisca replied:
We're filling in the address as well, so yes. I thought it'd be good to check that address fields also get persisted.

mikejolley commented:
@alexflorisca new test if failing, something about the batch. Left a comment.

Commit:
Update plugins/woocommerce-blocks/tests/e2e/tests/checkout/additional-fields.shopper.block_theme.spec.ts

Co-authored-by: Mike Jolley <mike.jolley@me.com>

Commit:
Remove check for batch request

Commit:
Fix CI failures

alexflorisca commented on plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/stock-indicator/test/block.test.tsx:
```diff
@@ -58,6 +58,10 @@ jest.mock( '@woocommerce/settings', () => ( {
 				simple: 'Simple product',
 			};
 		}
+		if ( param === 'globalPaymentMethods' ) {
```
`assets/js/settings/blocks/constants.ts` is loaded as part of this test (via a long stacktrace) now that we are importing constants to use in the `validateAdditionalFields` function. This test mocks `getSetting`, but it doesn't mock some of the keys used in the constants file, so the test fails. Added these extra mocks here

Commit:
Fix another waitForResponse

mikejolley reviewed (approved):
Did another run through testing instructions and I see CI is passing now so this should be ok to merge.

I would however suggest bulking out the testing instructions to include some additional fields so that other people not familiar with additonal field registration can test this PR.

Thanks!

alexflorisca commented:
Thanks Mike, edited the instructions with some code examples, good point!

Commit:
Fix flakey additional fields e2e tests and address being overriden

Dismiss modal

__internalIncrementCalculating when pushing checkout data

remove waitForCheckoutDataUpdate

Put back waits, and wait for customer data to update after filling in address fields

receiveCartContents instead of receiveCart so address is not replaced by response

receiveCartContent and increment

Change waitForCheckoutDataUpdate to use isCalculating() instead

Remove waits before place button

fillInCheckoutWithTestData should wait for billing or shipping to be visible

alexflorisca commented:
This commit https://github.com/woocommerce/woocommerce/pull/55989/commits/8c453683155848a1452b299c4adce891e531a3df is https://github.com/woocommerce/woocommerce/pull/56225 merged to this branch.

Commit:
Merge branch 'release/9.8' into fix/additional-fields-bugs

mikejolley committed:
Remove __internalIncrementCalculating when updating—we have cancellation in place

mikejolley reviewed (approved):
Overall I'm happy with the changes and I've gone through the testing steps a final time and it appears to be functional. I've tested that fields are no longer overwritten and watched console network logs to ensure thats working ok.

I think so long as CI passes (I pushed a small tweak) this should be good to ship. @senadir do you want to do a final check since this has grown in scope and we've collaborated on this?

senadir commented on plugins/woocommerce-blocks/assets/js/data/checkout/utils.ts:
```diff
@@ -230,3 +237,37 @@ export const getPaymentResultFromCheckoutResponse = (
+		} else if (
+			ORDER_FORM_KEYS.includes( fieldKey as keyof AdditionalValues )
+		) {
+			prefix = 'additional-information_';
```
I don't think this exists anymore, they should be prefixed with `order_`

senadir replied:
```suggestion
			prefix = 'order_';
```

senadir reviewed (approved):
I left a small about using the wrong error prefix but otherwise all is good.

WunderBart commented on plugins/woocommerce-blocks/tests/e2e/tests/checkout/additional-fields.shopper.block_theme.spec.ts:
```diff
@@ -908,5 +901,96 @@ test.describe( 'Shopper → Additional Checkout Fields', () => {
+			// Check that contact and order fields are persisted
+			await checkoutPageObject.page.waitForResponse(
+				'**/wc/store/v1/checkout**'
+			);
```
For stability, `waitForResponse` promise should be started before the actions that trigger the request are performed, e.g.:

```js
const responsePromise = this.page.waitForResponse( /* ... * / );

await this.page.getByLabel( 'Get it cheaper?' ).check();

await responsePromise;
```

More info here: https://playwright.dev/docs/api/class-page#page-wait-for-response

WunderBart commented on plugins/woocommerce-blocks/tests/e2e/tests/checkout/additional-fields.shopper.block_theme.spec.ts:
```diff
@@ -908,5 +901,96 @@ test.describe( 'Shopper → Additional Checkout Fields', () => {
+				.getByLabel( 'Test required checkbox' )
+				.check();
+
+			// This should generate a PUT request to checkout.
+			await checkoutPageObject.page.waitForResponse(
+				'**/wc/store/v1/checkout**'
+			);
```
Same as above.

WunderBart commented on plugins/woocommerce-blocks/tests/e2e/tests/checkout/checkout.page.ts:
```diff
@@ -68,6 +68,15 @@ export class CheckoutPage {
 			contact: {},
 		}
 	) {
+		await Promise.race( [
+			this.page
+				.getByRole( 'group', { name: 'Shipping address' } )
+				.waitFor( { state: 'visible' } ),
+			this.page
+				.getByRole( 'group', { name: 'Billing address' } )
+				.waitFor( { state: 'visible' } ),
+		] );
```
This will still throw a rejected promise in the trace, so let's use the `locator.or` here, e.g.:

```js
await this.page
	.getByRole( 'group', { name: 'Shipping address' } )
	.or( this.page.getByRole( 'group', { name: 'Billing address' } ) )
	.waitFor( { state: 'visible' } ),

```

More info here: https://playwright.dev/docs/api/class-locator#locator-or

alexflorisca replied:
I'm reluctant to do this because of this

> Note that when both locators match something, the resulting locator will have multiple matches, potentially causing a [locator strictness](https://playwright.dev/docs/locators#strictness) violation.

Both billing and shipping addresses may be visible, so this will introduce more flakiness potentially. Although I do get the improved stack trace this would give.

WunderBart replied:
In such a case, I think we can append `.first()` to that locator so we won't get yelled at by Playwright:

```js
await this.page
	.getByRole( 'group', { name: 'Shipping address' } )
	.or( this.page.getByRole( 'group', { name: 'Billing address' } ) )
	.first()
	.waitFor( { state: 'visible' } ),
```

mikejolley replied:
This got reopened yesterday, after this change I assume? https://github.com/woocommerce/woocommerce/issues/56037

WunderBart replied:
> This go reopened yesterday after this change I assume https://github.com/woocommerce/woocommerce/issues/56037

sorry, not sure I follow

alexflorisca replied:
Me either?

mikejolley replied:
The flakey test was reopened due to the tests in this PR. The problem with the `#email` selector. It only happened once—I wasn't sure if this logic was being changed and causing that to resurface.

Commit:
Validate additional fields on PUT

Commit:
Merge branch 'fix/additional-fields-bugs' of github.com:woocommerce/woocommerce into fix/additional-fields-bugs

Commit:
Update plugins/woocommerce-blocks/assets/js/data/checkout/utils.ts

Co-authored-by: Seghir Nadir <nadir.seghir@gmail.com>

Commit:
Improve waitForResponse

Commit:
Merge branch 'fix/additional-fields-bugs' of github.com:woocommerce/woocommerce into fix/additional-fields-bugs

alexflorisca commented:
@senadir and I noticed that additional fields were also not being validated on PUT today. This was happening because we missed the `validate_callback` of the PUT endpoint. Fixed in https://github.com/woocommerce/woocommerce/pull/55989/commits/95f80207f23b876db03eb61c0d89ab004d724fda. Nadir if you could give this a look over tomorrow and double check it works that would be great.

@WunderBart I addressed your comments about the `waitForResponse`, thanks for pointing that out. 

CI failures seem unrelated to this PR

Commit:
Replace Promise.race with  .or selector

Commit:
Merge branch 'release/9.8' into fix/additional-fields-bugs

alexflorisca commented:
Failing tests have been fixed on trunk. p1741956168053479/1741293577.344589-slack-C03CPM3UXDJ
